### PR TITLE
Rust rewrite

### LINKED ISSUE #54: Node crash when bundling (SourceMapBinding::addRawMappings)
Node crashes when parcel 2 (beta.2). I am currently trying to update a legacy app to use parcel, so I have some dependency within my repo (via bower).

Node Version 14.14.0 running on MacOS Big Sur 11.2.1 (20D74)

I set the `CI=true` env var in order to get a list:

```
- Building preprocess.html...

- Building main.preprocess.scss...

- Building can.js...

- Building can.jquery.js...

- Building header.js...

- Building category-slider.js...

- Building footer.js...

- Building html5-video-preview.js...

- Building movie-cover.js...

- Building isInViewport.js...

FATAL ERROR: Error::New napi_get_last_error_info
 1: 0x1012d84c5 node::Abort() (.cold.1) [$USER/.nvm/versions/node/v14.14.0/bin/node]
 2: 0x1000a5d59 node::Abort() [$USER/.nvm/versions/node/v14.14.0/bin/node]
 3: 0x1000a5ebf node::OnFatalError(char const*, char const*) [$USER/.nvm/versions/node/v14.14.0/bin/node]
 4: 0x1000a5d69 node::FatalError(char const*, char const*) [$USER/.nvm/versions/node/v14.14.0/bin/node]
 5: 0x10007bb4d napi_fatal_error [$USER/.nvm/versions/node/v14.14.0/bin/node]
 6: 0x10903c31a  [$PROJECT/node_modules/@parcel/source-map/prebuilds/darwin-x64/node.napi.glibc.node]
 7: 0x109041bc7 napi_register_module_v1 [$PROJECT/node_modules/@parcel/source-map/prebuilds/darwin-x64/node.napi.glibc.node]
 8: 0x10903cef4 SourceMapBinding::addRawMappings(Napi::CallbackInfo const&) [$PROJECT/node_modules/@parcel/source-map/prebuilds/darwin-x64/node.napi.glibc.node]
 9: 0x10903d1ff SourceMapBinding::_addNames(Napi::Array&) [$PROJECT/node_modules/@parcel/source-map/prebuilds/darwin-x64/node.napi.glibc.node]
10: 0x10903c9f9 SourceMapBinding::addRawMappings(Napi::CallbackInfo const&) [$PROJECT/node_modules/@parcel/source-map/prebuilds/darwin-x64/node.napi.glibc.node]
11: 0x109042922 Napi::InstanceWrap<SourceMapBinding>::InstanceVoidMethodCallbackWrapper(napi_env__*, napi_callback_info__*) [$PROJECT/node_modules/@parcel/source-map/prebuilds/darwin-x64/node.napi.glibc.node]
12: 0x10904289a Napi::InstanceWrap<SourceMapBinding>::InstanceVoidMethodCallbackWrapper(napi_env__*, napi_callback_info__*) [$PROJECT/node_modules/@parcel/source-map/prebuilds/darwin-x64/node.napi.glibc.node]
13: 0x10006373a v8impl::(anonymous namespace)::FunctionCallbackWrapper::Invoke(v8::FunctionCallbackInfo<v8::Value> const&) [$USER/.nvm/versions/node/v14.14.0/bin/node]
14: 0x100258938 v8::internal::FunctionCallbackArguments::Call(v8::internal::CallHandlerInfo) [$USER/.nvm/versions/node/v14.14.0/bin/node]
15: 0x100257ecc v8::internal::MaybeHandle<v8::internal::Object> v8::internal::(anonymous namespace)::HandleApiCallHelper<false>(v8::internal::Isolate*, v8::internal::Handle<v8::internal::HeapObject>, v8::internal::Handle<v8::internal::HeapObject>, v8::internal::Handle<v8::internal::FunctionTemplateInfo>, v8::internal::Handle<v8::internal::Object>, v8::internal::BuiltinArguments) [$USER/.nvm/versions/node/v14.14.0/bin/node]
16: 0x100257632 v8::internal::Builtin_Impl_HandleApiCall(v8::internal::BuiltinArguments, v8::internal::Isolate*) [$USER/.nvm/versions/node/v14.14.0/bin/node]
17: 0x100a70a99 Builtins_CEntry_Return1_DontSaveFPRegs_ArgvOnStack_BuiltinExit [$USER/.nvm/versions/node/v14.14.0/bin/node]
18: 0x100a09d62 Builtins_InterpreterEntryTrampoline [$USER/.nvm/versions/node/v14.14.0/bin/node]

```

Maybe there is something in [isInViewport](https://github.com/zeusdeux/isInViewport) that breaks it, but removing the import will lead to it crashing later. More specifically, rerunning it will lead to it crashing at different times/files.

Unfortunately, I cannot provide a minimal reproduction example. But I could add somebody to the private repository where the error occurs.

### PR DESCRIPTION
**Pull request by DeMoorJasper**
This PR is a ground-up rewrite of this library in Rust instead of C++.
This should result in a more stable and easier to read codebase, with the advantage of having access to the vast amount of libraries that are already available for Rust.

Fixes #54

TODO:
- [x] Setup benchmark action and compare to current version
- [x] Fix failing tests
- [x] Update azure pipeline
- [x] Move filepath normalization to Rust
- [x] Fix performance issues with filepath normalization in Rust
- [x] Experiment with bincode instead of flatbuffers
- [x] Implement extends
- [x] Setup test workflow
- [x] Setup release workflow

### CODE REVIEW

**Comment by devongovett** [comment]
Nice! What do you think about eventually putting this in the monorepo with the other rust packages? Would unify the CI stuff at least...

**Comment by DeMoorJasper** [comment]
@devongovett yes we should probably make this part of the monorepo at some point. Although it would make it even slower... maybe we should split the ci up into multiple github actions so they run in parallel?

**Comment by DeMoorJasper** [comment]
Can possible use this instead of bincode: https://crates.io/crates/flexbuffers

**Comment by devongovett** [comment]
Should we use the same rustfmt settings as in the monorepo?

**Comment by DeMoorJasper** [comment]
@devongovett Sure, feel free to update it

File: parcel_sourcemap_node/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,581 @@
...
+
+    let source = ctx.get::<JsString>(0)?.into_utf8()?;
+    let source_index = source_map_instance.get_source_index(source.as_str()?)?;
+    match source_index {
+        Some(i) => {
+            let source_content = source_map_instance.get_source_content(i)?;
+            return ctx.env.create_string(source_content);
+        }
+        None => {
+            return ctx.env.create_string("");
```
Comment (devongovett) [inline parcel_sourcemap_node/src/lib.rs:126]: Should this throw an error or return null maybe?

File: parcel_sourcemap_node/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,581 @@
...
+    let this: JsObject = ctx.this_unchecked();
+    let source_map_instance: &SourceMap = ctx.env.unwrap(&this)?;
+
+    let name_index = ctx.get::<JsNumber>(0)?.get_uint32()?;
+    match source_map_instance.get_name(name_index) {
+        Ok(name) => {
+            return ctx.env.create_string(name);
+        }
+        Err(_err) => {
+            return ctx.env.create_string("");
```
Comment (devongovett) [inline parcel_sourcemap_node/src/lib.rs:152]: Same with this one. I guess this is ignoring the error atm

File: parcel_sourcemap_node/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,581 @@
...
+
+    return Ok(result_obj);
+}
+
+#[js_function(1)]
+fn add_indexed_mappings(ctx: CallContext) -> Result<JsUndefined> {
+    let this: JsObject = ctx.this_unchecked();
+    let source_map_instance: &mut SourceMap = ctx.env.unwrap(&this)?;
+
+    // TODO: Figure out a more optimal way of handling typed arrays...
```
Comment (devongovett) [inline parcel_sourcemap_node/src/lib.rs:365]: I believe there is `JsTypedArray`/`JsArrayBuffer`: https://github.com/napi-rs/napi-rs/blob/e4753b7c1d839ab53de42107aeb5395db8749431/napi/src/js_values/arraybuffer.rs.

There's also `JsBuffer` if we wanna use Node buffers. https://napi.rs/concepts/js-values#jsbuffer

File: src/SourceMap.js [hunk 4]
```diff
@@ -13,15 +14,18 @@ export default class SourceMap {
   /**
    * @private
    */
-  projectRoot: string;
+  projectRoot: any;
```
Comment (devongovett) [inline src/SourceMap.js:17]: ?

File: test/find.test.js [hunk 5]
```diff
@@ -12,30 +12,30 @@ describe('SourceMap - Find', () => {
 
     let mapping = map.findClosestMapping(2, 8);
     assert.deepEqual(mapping, {
-      generated: { line: 2, column: 9 },
-      original: { line: 1, column: 7 },
-      source: './helloworld.coffee',
+      generated: { line: 2, column: 2 },
+      original: { line: 1, column: 0 },
+      source: 'helloworld.coffee',
```
Comment (devongovett) [inline test/find.test.js:17]: This change was intentional?

**Review by devongovett — APPROVED** [review]
Looks good! A few comments but nothing blocking. Should we also update the readme to say it's written in Rust instead of C++ now?

**Comment by DeMoorJasper** [comment]
@devongovett Thanks, will update readme and resolve your comments

File: parcel_sourcemap_node/src/lib.rs (hunk 1 above)
Comment (DeMoorJasper) [reply parcel_sourcemap_node/src/lib.rs:126]: This is intentional to mimic the c++ logic.  Might make some breaking changes in a follow-up

File: parcel_sourcemap_node/src/lib.rs (hunk 2 above)
Comment (DeMoorJasper) [reply parcel_sourcemap_node/src/lib.rs:152]: Same consistency w/ cpp implementation

File: src/SourceMap.js (hunk 4 above)
Comment (DeMoorJasper) [reply src/SourceMap.js:17]: probably from some experimenting and flow complaining, reverted that...

File: test/find.test.js (hunk 5 above)
Comment (DeMoorJasper) [reply test/find.test.js:17]: Yes, wanted to make it follow the same logic as `source-map` library, esbuild, ...

Comment (DeMoorJasper) [reply test/find.test.js:17]: I can revert it very easily

File: parcel_sourcemap_node/src/lib.rs (hunk 3 above)
Comment (DeMoorJasper) [reply parcel_sourcemap_node/src/lib.rs:365]: Yeah been discussing how to properly tackle this with the napi-rs maintainer, gonna update that... it's really hitting performance atm

**Pull request merged**